### PR TITLE
Port to ClojureScript self-host

### PR DESCRIPTION
**Pull request by arichiardi**
This allows to use cljs-tooling in lumo, planck and in general ClojureScript
self-host environments.

### CODE REVIEW

**Comment by arichiardi** [comment]
ClojureScript vanilla tests should pass again.

**Comment by arichiardi** [comment]
This is becoming more of a rewrite/refactoring to make things compatible for both JVM and self-host. Still things to do, not in a rush :smile:

**Comment by bbatsov** [comment]
That is fine.

**Comment by arichiardi** [comment]
All green, both `lumo` and `jvm` tests pass :tada:

File: deps.edn [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{:paths ["src"]
```
Comment (bbatsov) [inline deps.edn]: Why is this needed?

File: scripts/test-jvm [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env sh
+
+set -eu
+
+clojure -Srepro -A:test:runner:cljs-jvm -r '^cljs-tooling.*' $@
```
Comment (bbatsov) [inline scripts/test-jvm]: I'm a bit puzzled that some scripts use lein, some tools.deps. Some explanations would be useful.

Comment (arichiardi) [reply scripts/test-jvm]: I found that `lein test` was not working on my machine, it seemed to do some caching. I had `clojure` scripts handy and decided to use those. It will also nice down the road to use this project as git dep.

File: deps.edn (hunk 1 above)
Comment (arichiardi) [reply deps.edn]: Not needed indeed

File: scripts/test-jvm (hunk 2 above)
Comment (arichiardi) [reply scripts/test-jvm]: tried again and I get

```
Exception in thread "main" java.lang.RuntimeException: No such var: u/add-ns-macros, compiling:(cljs_tooling/test_util.cljc:7:15)
...
```

weird..I don't know what is going on here...

File: src/cljs_tooling/util/analysis.cljc [hunk 3]
```diff
@@ -137,10 +137,29 @@
...
+    :req :req-un :opt :opt-un
+    :args :ret :fn
+    :const
+    :arglists :tag :static :added})
+
 (defn keyword-constants
   "Returns a list of keyword constants in the environment."
   [env]
-  (filter keyword? (keys (:cljs.analyzer/constant-table env))))
+  (concat common-keywords (filter keyword? (keys (:cljs.analyzer/constant-table env)))))
```
Comment (bbatsov) [inline src/cljs_tooling/util/analysis.cljc]: Why isn't the analyzer returning those? Is tis some bug or something?

File: src/cljs_tooling/util/analysis.cljc [hunk 4]
```diff
@@ -137,10 +137,29 @@
         excludes (:excludes (find-ns env ns))]
     (apply dissoc macros excludes)))
 
+(def ^:private common-keywords
```
Comment (bbatsov) [inline src/cljs_tooling/util/analysis.cljc]: Not sure that `common` is the right term here. Not sure how exactly did you come up with the list - might be a good idea to explain this in the docstring.

File: .travis.yml [hunk 5]
```diff
@@ -1,16 +1,21 @@
...
+script:
+  - ./scripts/test-jvm
+  - ./scripts/test-lumo
 branches:
   only:
     - master
 jdk:
-  - openjdk7
-  - oraclejdk7
+  - openjdk8
```
Comment (bbatsov) [inline .travis.yml]: Seems to me you should add here JDK 9 and 10 as well.

File: src/cljs_tooling/util/analysis.cljc (hunk 3 above)
Comment (arichiardi) [reply src/cljs_tooling/util/analysis.cljc]: I'll update the docstring, about the why...well this is a question for David but I guess these keywords are not part of the compiler state, are just used and then discarded...

I took the list from lumo, which is in any case arbitrary.

File: src/cljs_tooling/util/analysis.cljc (hunk 4 above)
Comment (arichiardi) [reply src/cljs_tooling/util/analysis.cljc]: From `lumo`, and it is arbitrary. I guess in Clojure there is a similar list somewhere?

File: scripts/test-jvm (hunk 2 above)
Comment (arichiardi) [reply scripts/test-jvm]: While I agree with you for consistency, for CI the new cli tool is a better tool - faster and resolves deps more precisely. Will try to debug but I cannot promise I can spend a lot of time on this...all tests are green :))))

File: src/cljs_tooling/util/analysis.cljc [hunk 6]
```diff
@@ -1,5 +1,7 @@
 (ns cljs-tooling.util.analysis
-  (:require [cljs-tooling.util.misc :as u])
+  (:require [cljs-tooling.util.misc :as u]
+            #?(:clj [cljs.repl :as repl]
+               :cljs [lumo.repl-resources :as repl]))
```
[the remark this answers is not in the document]
Comment (arichiardi) [reply src/cljs_tooling/util/analysis.cljc]: Right, I might need dynamic requires, which I doubt will work. Will try.

File: .travis.yml (hunk 5 above)
Comment (arichiardi) [reply .travis.yml]: Can do that

File: src/cljs_tooling/util/analysis.cljc (hunk 6 above)
[the remark this answers is not in the document]
Comment (arichiardi) [reply src/cljs_tooling/util/analysis.cljc]: Opened a JIRA on this https://dev.clojure.org/jira/browse/CLJS-2808

File: src/cljs_tooling/util/analysis.cljc (hunk 3 above)
Comment (bbatsov) [reply src/cljs_tooling/util/analysis.cljc]: Well, I guess someone should ask him what's the proper way to get those, so we don't have to hardcode them. :-)

File: src/cljs_tooling/util/analysis.cljc (hunk 4 above)
Comment (bbatsov) [reply src/cljs_tooling/util/analysis.cljc]: Yeah, I just think it'd be better to use this list directly, instead of hardcoding something ourselves, but that's not a big deal.

File: src/cljs_tooling/util/analysis.cljc (hunk 6 above)
[the remark this answers is not in the document]
Comment (bbatsov) [reply src/cljs_tooling/util/analysis.cljc]: 👍

**Comment by bbatsov** [comment]
OK, let me know when I should review this again.

**Comment by arichiardi** [comment]
Yep sorry I started doing things that got interrupted, will mention when ready, thanks a lot ;)

**Comment by bbatsov** [comment]
And a reminder to wrap this up as well. :-)

**Comment by bbatsov** [comment]
Sounds good to me. Let's split this in parts.

**Comment by arichiardi** [comment]
Another solution would be to copy those two maps over to this project - they should be relatively stable   for us to maintain them in sync.

**Comment by arichiardi** [comment]
I opted for including the namespace here. I think that if tests are good, we can review this again ;)

**Comment by arichiardi** [comment]
@bbatsov I got rid of `clojure` because of https://dev.clojure.org/jira/browse/TDEPS-20 - maybe what was causing the problems with JDK > 9.

Now `lein` drives all the scripts and if tests pass this is now good to go.

**Comment by arichiardi** [comment]
Alex actually directed me to https://github.com/travis-ci/travis-ci/issues/9368. Which is probably what is happening.

**Comment by arichiardi** [comment]
I think I solved it, thanks also to @puredanger's very useful hint

**Comment by bbatsov** [comment]
New snapshot is out.

**Comment by bbatsov** [comment]
This also seems relevant https://www.deps.co/guides/travis-ci-latest-java/

**Comment by arichiardi** [comment]
Awesome, yes I think she post is basically what I ended up doing. Will try this out in my local nrepl
